spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -790,4 +802,64 @@ public Identifier identifier() {
   public static TableIdentifier identifierToTableIdentifier(Identifier identifier) {
     return TableIdentifier.of(Namespace.of(identifier.namespace()), identifier.name());
   }
+
+  /**
+   * Use Spark to list all partitions in the table.
+   *
+   * @param spark a Spark session
+   * @param rootPath a table identifier
+   * @param format format of the file
+   * @return all table's partitions
+   */
+  public static List<SparkPartition> getPartitions(SparkSession spark, Path rootPath, String format) {
+    FileStatusCache fileStatusCache = FileStatusCache.getOrCreate(spark);
+    Map<String, String> emptyMap = Collections.emptyMap();
+
+    InMemoryFileIndex fileIndex = new InMemoryFileIndex(
+        spark,
+        JavaConverters
+            .collectionAsScalaIterableConverter(ImmutableList.of(rootPath))
+            .asScala()
+            .toSeq(),
+        JavaConverters
+            .mapAsScalaMapConverter(emptyMap)
+            .asScala()
+            .toMap(Predef.<Tuple2<String, String>>conforms()),
+        Option.empty(),
+        fileStatusCache,
+        Option.empty(),
+        Option.empty());
+
+    org.apache.spark.sql.execution.datasources.PartitionSpec spec = fileIndex.partitionSpec();
+    StructType schema = spec.partitionColumns();
+    if (spec.partitions().isEmpty()) {
+      return ImmutableList.of(new SparkPartition(Collections.emptyMap(), rootPath.toString(), format));
```

Comment:
sure we can do that, I think that's also reasonable

Reply:
hmm this does mean we don't check whether we are importing from a partitioned source .... let me think about this a bit more

Reply:
Well, if we want to validate that explicitly, we can do something like this.

```
if (table.spec().isUnpartitioned()) {
  List<SparkPartition> partitions = Spark3Util.getPartitions(spark(), tableLocation, format);
  Preconditions.checkArgument(partitions.size() == 0, ...);
  
  SparkPartition partition = new SparkPartition(Collections.emptyMap(), tableLocation.toString(), format);
  importPartitions(table, ImmutableList.of(partition));
} else {
  List<SparkPartition> partitions = Spark3Util.getPartitions(spark(), tableLocation, format);
  Preconditions.checkArgument(partitions.size() > 0, ...);

  List<SparkPartition> filteredPartitions = SparkTableUtil.filterPartitions(partitions, partitionFilter);
  Preconditions.checkArgument(...);

  importPartitions(table, filteredPartitions);
}
```

I am just thinking out loud here. The validation logic in `importFileTable` was a bit hard to follow.

Reply:
Up to you, @RussellSpitzer.

Reply:
The issue here is that getPartitions throws an exception here
```
      Seq<CatalogTablePartition> partitions = catalog.listPartitions(tableIdent, Option.empty());
```
If there are no partitions, but we could change the code so that it returns an empy seq if the table in unpartitioned

Reply:
That ... may have been a different problem ... I'll figure this out